fcrepo-persistence-api/src/main/java/org/fcrepo/persistence/api/PersistentStorageSession.java

```diff
@@ -0,0 +1,114 @@
+ * limitations under the License.
+ */
+package org.fcrepo.persistence.api;
+
+import java.time.Instant;
+
+import org.fcrepo.kernel.api.exception.CannotCreateResourceException;
+import org.fcrepo.kernel.api.exception.ItemNotFoundException;
+import org.fcrepo.kernel.api.exception.RepositoryRuntimeException;
+import org.fcrepo.kernel.api.models.FedoraResource;
+import org.fcrepo.kernel.api.models.Tombstone;
+
+/**
+ * An abstract class that mediates CRUD operations to and from persistence storage.
+ *
+ * @author dbernstein
+ * @author whikloj
+ */
+abstract public class PersistentStorageSession {
+
+    /**
+     * The FedoraTransaction ID.
+     */
+    private final String txId;
+
+    /**
+     * Basic constructor.
+     *
+     * @param txId the FedoraTransaction ID.
+     */
+    protected PersistentStorageSession(final String txId) {
+        this.txId = txId;
+    }
+
+    /**
+     * Create a new resource on the persistent storage.
+     *
+     * @param resource The new fedora resource to persist to storage.
+     * @return the resource with any new information related to the persisting..?
+     * @throws CannotCreateResourceException Error persisting the resource.
+     */
+    abstract public FedoraResource create(final FedoraResource resource)
+            throws CannotCreateResourceException;
+
+    /**
+     * Update an existing resource on persistent storage.
+     *
+     * @param resource The changed fedora resource to persist to storage.
+     * @return the updated resource with any new information related to the persisting..?
+     * @throws CannotCreateResourceException Error persisting the resource.
+     */
+    abstract public FedoraResource update(final FedoraResource resource)
+            throws CannotCreateResourceException;
+
+    /**
+     * Delete a resource from persistent storage.
+     *
+     * @param resource The current fedora resource to delete.
+     * @return The tombstone for the removed resource.
+     * @throws RepositoryRuntimeException Error deleting the resource.
+     */
+    abstract public Tombstone delete(final FedoraResource resource)
+            throws RepositoryRuntimeException;
+
+    /**
+     * Return a resource from persistent storage
+     *
+     * @param identifier The identifier of the resource to retrieve.
+     * @return The resource.
+     * @throws ItemNotFoundException If the identifier doesn't exist.
+     */
+    abstract public FedoraResource read(final String identifier)
+            throws ItemNotFoundException;
+
+    /**
+     * Return a version of a resource from persistent storage
+     *
+     * @param identifier The identifier of the resource to retrieve.
+     * @param version The date/time of the version to retrieve.
+     * @return The version of the resource.
+     * @throws ItemNotFoundException If the identifier doesn't exist.
+     */
+    abstract public FedoraResource read(final String identifier, final Instant version) throws ItemNotFoundException;
```

Comment:
Do we want to have to create a session in order to perform a read operation against immutable state?

Reply:
Perhaps the PersistentStorageSessionFactory could provide a method like this: 
`ReadOnlyPersistentStorageSession createReadOnlySession()`
and ReadOnlyPersistentStorageSession would have a single method: 
`FedoraResource read(final String identifier, final Instant version) throws ItemNotFoundException` ?